src/core/expression/qgsexpressionfunction.cpp

```diff
@@ -4930,6 +4930,12 @@ static QVariant executeGeomOverlay( const QVariantList &values, const QgsExpress
     request.setFilterExpression( filterString ); //filter cached features
   }
 
+  if (targetLayer->crs().authid() != sourceLayer->crs().authid())
+  {
+      QgsCoordinateTransformContext TransformContext;
+      request.setDestinationCrs(sourceLayer->crs(), TransformContext); //if crs are not the same, cached target will be reprojected to source crs
```

Comment:
A new project for self overlay and different crs test

[overlay_test5.zip](https://github.com/m-kuhn/QGIS/files/3391013/overlay_test5.zip)